### PR TITLE
fix(ui): re-anchor scroll on live prepends in top-anchored virtual lists

### PR DESCRIPTION
**Pull request by panghy**
Follow-up to #158, addressing the review thread on scroll position during live merges (https://github.com/endara-ai/endara-desktop/pull/158#discussion_r3873006736).

## Problem

The Observability call list renders through `VirtualLogList` in top-anchored mode (`followTail={false}`). Live refresh merges incoming first-page rows via `mergeCalls`, which sorts newest-first — so new rows **prepend**. Nothing compensated `scrollTop`, and the absolutely/transform-positioned virtual rows defeat native browser scroll anchoring (`overflow-anchor`) that the old plain table benefited from on Chromium-based WebView2. Result: a user scrolled mid-list reading older calls saw the content under the cursor shift down by the prepended height on every terminal-event burst (most visible on Windows; macOS WKWebView never had native anchoring).

## Fix

Re-anchor around item keys when `followTail === false` and `scrollTop > 0`:

- **Capture** (`$effect.pre`, runs before the DOM updates and before the count-sync effect, so the virtualizer still reflects the old layout): record the first visible row's key and its offset relative to the viewport top (`captureScrollAnchor`).
- **Restore** (after `tick()`): find the anchor key's new index (`findIndexByKey`), read its new start offset from the virtualizer's measurement cache, and scroll so the same row sits at the same viewport offset (`restoredScrollTop`).

Measured row heights are keyed by item key in `@tanstack/virtual-core`, so the anchor's new offset is consistent before/after a prepend and the correction is exact for the visible region.

No-op by construction when:

- `followTail` is true — tail-pinned views (LogsTab, RelayLogs) are completely unaffected; capture never happens.
- `scrollTop === 0` — the user is reading the newest rows, where prepends should push content down.
- the merge only appends (Load more) — the anchor's offset is unchanged, so the computed target equals the current `scrollTop`.
- the anchor row was evicted by the `LIVE_MAX_ROWS` cap from #157 — no correction rather than a guess.

The decision logic lives in `virtual-log-list-helpers.ts` as pure functions, per the existing extracted-helpers + Node-env vitest pattern. No data-layer changes (`mergeCalls`, cursors, `LIVE_MAX_ROWS` untouched).

## Testing

- `pnpm test`: 1229 passed, 39 skipped — includes new suites for `captureScrollAnchor` (no capture in tail-follow mode / at top / with no rows; skips overscan rows fully above the viewport; exact viewport-offset capture), `findIndexByKey` (shifted index after prepend, eviction → -1), and `restoredScrollTop` (offset restore, append no-op, eviction no-op, top clamp).
- `pnpm check`: 0 errors (1 pre-existing `tsconfig` 'node' types warning, baseline).

### CODE REVIEW

File: src/lib/components/VirtualLogList.svelte [hunk 1]
```diff
@@ -79,6 +83,28 @@
...
+  // the transform-positioned virtual rows defeat native scroll anchoring, so
+  // content under the cursor would shift down on every merge. Before the new
+  // items render ($effect.pre runs ahead of DOM updates and the count-sync
+  // effect below, so getVirtualItems() still reflects the OLD layout),
+  // remember the first visible row's key + viewport offset. Tail-follow mode
+  // and scrollTop 0 capture nothing (see captureScrollAnchor).
+  let pendingAnchor: ScrollAnchor | null = null;
+  $effect.pre(() => {
+    void items.length;
+    const follow = followTail;
```
[the remark this answers is not in the document]
Comment (panghy) [reply src/lib/components/VirtualLogList.svelte:104]: Verified — this is not a bug in practice, though the comment could be clearer about why.

In Svelte 5 runes mode, reading `items.length` inside the effect first reads the `items` prop signal itself, then `.length` on it. Every live merge reassigns the parent state to a **new array** (`mergeCalls` always builds a fresh array via `[...byUid.values()].sort(...)`, including the capped `merged.slice(0, maxRows)` path), so the prop reference changes and both anchor effects re-run even when the length stays pinned at `LIVE_MAX_ROWS`. I confirmed the at-cap path: `calls = mergeCalls(calls, callsRes.calls, Math.max(LIVE_MAX_ROWS, calls.length))` in `Observability.svelte` cannot mutate in place.

Non-blocking suggestion: change `void items.length` to `void items` (or note the reference-tracking intent in the comment) in both effects, so the dependency doesn't read as "length changes only".

**Review by panghy — COMMENTED** [review]
## Review: APPROVE (posted as comment — GitHub blocks self-approval by the PR author's token)

Verified the capture/restore design against Svelte 5 effect semantics and the installed `@tanstack/virtual-core@3.17.7` internals, plus the four consumer edge cases. No blocking issues found. `vitest run`: 52 files / 1229 tests pass locally.

### Correctness checks performed

1. **Capture sees the OLD layout in `$effect.pre`** — verified. `getVirtualItems()` is memoized on `[getVirtualIndexes(), getMeasurements()]`, and `getMeasurements()` on `[options.count, …, options.getItemKey, itemSizeCacheVersion]`. Since the count-sync `setOptions` runs in a later regular `$effect` and `getItemKey` is a stable closure, the memo hits cache during `$effect.pre` and returns pre-merge keys/offsets. (Narrow caveat: if a row resize lands in the same flush as the items change, `itemSizeCacheVersion` busts the memo and measurements recompute with new-array keys at the old count — worst case is one mildly imperfect restore, self-limiting. Not blocking.)

2. **At-cap merges (`LIVE_MAX_ROWS`) still trigger capture/restore** — Copilot flagged that `void items.length` misses same-length content shifts. Not an actual bug: `mergeCalls` always returns a new array, and in Svelte 5 reading `items.length` tracks the `items` prop reference itself, which changes on every merge. Replied on the thread; a one-word comment tweak (see thread) would make the intent explicit.

3. **Restore math vs. measurement cache** — `measurementsCache[index]?.start` is valid in v3.17.7's `lanes === 1` fast path (`createLazyMeasurementsView` over the flat array), and item sizes are cached **by key**, so previously measured rows keep real heights across the prepend; only the new rows use the 20 px estimate. When those later mount and measure, virtual-core's first-measure adjustment (`itemStart < scrollOffset` ⇒ `applyScrollAdjustment(delta)`) compensates the estimate→actual delta, so there is no compounding drift after the restore.

4. **Stale-anchor safety** — the part-1 `$effect.pre` dependency set (`items`, `followTail`, `scrollEl`, `virtualizer`) is a superset of part 2's (`items`, `virtualizer`) and pre-effects run first, so any anchor left over from a `followTail`/`scrollEl`-only rerun is overwritten before part 2 consumes it. `pendingAnchor` is consumed exactly once.

5. **Edge cases** — `followTail=true` (LogsTab/RelayLogs): `captureScrollAnchor` returns `null` unconditionally, restore no-ops; the only added work is one memo-cached `getVirtualItems()` read per items change. `scrollTop 0`: no capture (correct for a newest-first list — the user is reading the live head). "Load more" append: anchor start unchanged ⇒ `restoredScrollTop` returns `null`. Anchor evicted by the cap: `findIndexByKey` ⇒ −1 ⇒ no-op. All four are unit-tested.

6. **Timing** — the restore runs in `tick().then()`, i.e. after the DOM update but before paint, so the corrected scrollTop lands in the same frame as the prepended rows (no flicker).

### Tests & style
Helper tests follow the folder's established pure-function pattern (Node env, no jsdom) and cover the meaningful branches including the clamp-at-0 and already-at-0 no-op. Comment density and structure match the surrounding code.

### Non-blocking suggestions
- Consider `void items;` (or a comment noting the reference-tracking intent) instead of `void items.length` in the two anchor effects — the correctness at the `LIVE_MAX_ROWS` cap depends on the array *reference* changing, and `length` alone reads as if same-length merges were intentionally excluded.

File: src/lib/components/VirtualLogList.svelte (hunk 1 above)
[the remark this answers is not in the document]
Comment (panghy) [reply src/lib/components/VirtualLogList.svelte:104]: The effect tracks the `items` prop signal itself, not just the numeric length — in Svelte 5, `items.length` first reads the prop, so any reference replacement reruns the effect even when the length is unchanged. The capped live-refresh scenario goes through `mergeCalls`, which always returns a new array, so capture does run there. What genuinely isn't tracked is in-place mutation of the same array — but that already breaks the rest of the component (the count sync and the virtualizer's getItemKey diffing assume immutable replacement too). Documented that contract on the `items` prop in 05a63d0.

File: src/lib/components/VirtualLogList.svelte [hunk 2]
```diff
@@ -79,6 +89,30 @@
...
+    const el = scrollEl;
+    if (!el) {
+      pendingAnchor = null;
+      return;
+    }
+    const rows = get(virtualizer)
+      .getVirtualItems()
+      .map((v) => ({ key: String(v.key), start: v.start, end: v.end }));
+    pendingAnchor = captureScrollAnchor(follow, el.scrollTop, rows);
+  });
```
[the remark this answers is not in the document]
Comment (panghy) [reply src/lib/components/VirtualLogList.svelte:114]: Addressed in #160: the capture effect now bails (and clears `pendingAnchor`) before calling `getVirtualItems()` when `followTail` is true or `scrollTop <= 0`, so tail-pinned log views skip the work entirely. `captureScrollAnchor` keeps its own guards so the helper contract is unchanged.

File: src/lib/components/VirtualLogList.svelte [hunk 3]
```diff
@@ -87,6 +121,30 @@
...
+      if (!el) return;
+      const inst = get(store);
+      // Refresh the measurement cache before reading row offsets.
+      inst.getVirtualItems();
+      const index = findIndexByKey(items, anchor.key, getKey);
+      const newStart = index === -1 ? null : (inst.measurementsCache[index]?.start ?? null);
+      const target = restoredScrollTop(anchor, newStart, el.scrollTop);
+      if (target !== null) inst.scrollToOffset(target);
+    });
+  });
```
[the remark this answers is not in the document]
Comment (panghy) [reply src/lib/components/VirtualLogList.svelte:146]: Addressed in #160: the restore effect now snapshots `items` and `getKey` before scheduling the `tick()` callback, and the callback bails if `items` has been replaced since — the newer merge's own effect run owns the restore in that case, so a stale anchor can never be applied against a newer array.
